src/server/qgswmsserver.cpp

```diff
@@ -984,36 +983,33 @@ void QgsWmsServer::runHitTest( QPainter* painter, HitTest& hitTest )
 {
   QPaintDevice* thePaintDevice = painter->device();
 
-  // setup QgsRenderContext in the same way as QgsMapRenderer does
+  // setup QgsRenderContext in the same way as the renderer does
   QgsRenderContext context;
   context.setPainter( painter ); // we are not going to draw anything, but we still need a working painter
   context.setRenderingStopped( false );
-  context.setRasterScaleFactor(( thePaintDevice->logicalDpiX() + thePaintDevice->logicalDpiY() ) / 2.0 / mMapRenderer->outputDpi() );
-  context.setScaleFactor( mMapRenderer->outputUnits() == QgsMapRenderer::Millimeters ? mMapRenderer->outputDpi() / 25.4 : 1.0 );
-  context.setRendererScale( mMapRenderer->scale() );
-  context.setMapToPixel( *mMapRenderer->coordinateTransform() );
-  context.setExtent( mMapRenderer->extent() );
+  context.setRasterScaleFactor(( thePaintDevice->logicalDpiX() + thePaintDevice->logicalDpiY() ) / 2.0 / mMapSettings->outputDpi() );
+  context.setScaleFactor( mConfigParser->outputUnits() == QgsUnitTypes::RenderUnit::RenderMillimeters ? mMapSettings->outputDpi() / 25.4 : 1.0 );
+  context.setRendererScale( mMapSettings->scale() );
+  context.setMapToPixel( mMapSettings->mapToPixel() );
+  context.setExtent( mMapSettings->extent() );
```

Comment:
I think this should use QgsRenderContext::fromMapSettings initially, and then override the things it needs for server. It's less maintenance/more future proof that way.